File: mobile/android/FUN/app/src/main/java/com/fun/app/ui/viewmodels/AuthViewModel.kt
/**
 * FUN App - Auth ViewModel
 */

package com.fun.app.ui.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.fun.app.FunApplication
import com.fun.app.core.auth.AuthManager
import com.fun.app.core.network.ApiClient
import com.fun.app.core.network.NetworkResult
import com.fun.app.core.network.safeApiCall
import com.fun.app.data.models.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class AuthViewModel(
    private val apiClient: ApiClient,
    private val authManager: AuthManager
) : ViewModel() {

    private val _isAuthenticated = MutableStateFlow(authManager.isAuthenticated.value)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()

    private val _currentUser = MutableStateFlow<User?>(null)
    val currentUser: StateFlow<User?> = _currentUser.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    init {
        viewModelScope.launch {
            authManager.isAuthenticated.collect { isAuth ->
                _isAuthenticated.value = isAuth
                if (isAuth && _currentUser.value == null) {
                    fetchProfile()
                }
            }
        }
    }

    fun signup(email: String, password: String, displayName: String) {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null

            when (val result = safeApiCall {
                apiClient.apiService.signup(
                    SignupRequest(email, password, displayName)
                )
            }) {
                is NetworkResult.Success -> {
                    authManager.saveTokens(
                        accessToken = result.data.tokens.accessToken,
                        refreshToken = result.data.tokens.refreshToken
                    )
                    _currentUser.value = result.data.user
                }
                is NetworkResult.Error -> {
                    _errorMessage.value = result.message
                }
                else -> {}
            }

            _isLoading.value = false
        }
    }

    fun login(email: String, password: String) {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null

            when (val result = safeApiCall {
                apiClient.apiService.login(LoginRequest(email, password))
            }) {
                is NetworkResult.Success -> {
                    authManager.saveTokens(
                        accessToken = result.data.tokens.accessToken,
                        refreshToken = result.data.tokens.refreshToken
                    )
                    _currentUser.value = result.data.user
                }
                is NetworkResult.Error -> {
                    _errorMessage.value = result.message
                }
                else -> {}
            }

            _isLoading.value = false
        }
    }

    fun logout() {
        authManager.clearTokens()
        _currentUser.value = null
    }

    private fun fetchProfile() {
        viewModelScope.launch {
            when (val result = safeApiCall { apiClient.apiService.getProfile() }) {
                is NetworkResult.Success -> {
                    _currentUser.value = result.data.user
                }
                is NetworkResult.Error -> {
                    if (result.code == 401) {
                        logout()
                    }
                }
                else -> {}
            }
        }
    }

    fun updateProfile(displayName: String?, avatarUrl: String?, onSuccess: () -> Unit) {
        viewModelScope.launch {
            _isLoading.value = true

            when (val result = safeApiCall {
                apiClient.apiService.updateProfile(
                    UpdateProfileRequest(displayName, avatarUrl)
                )
            }) {
                is NetworkResult.Success -> {
                    _currentUser.value = result.data.user
                    onSuccess()
                }
                is NetworkResult.Error -> {
                    _errorMessage.value = result.message
                }
                else -> {}
            }

            _isLoading.value = false
        }
    }

    companion object {
        val Factory: ViewModelProvider.Factory = object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>): T {
                val app = FunApplication.instance
                return AuthViewModel(
                    apiClient = app.apiClient,
                    authManager = app.authManager
                ) as T
            }
        }
    }
}

File: mobile/ios/FUN/FUN/ViewModels/AuthViewModel.swift
/**
 * FUN App - Auth ViewModel
 * Handles authentication state and operations
 */

import Foundation
import Combine

@MainActor
class AuthViewModel: ObservableObject {
    @Published var isAuthenticated = false
    @Published var currentUser: User?
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    private let authManager = AuthManager.shared
    private let apiClient = APIClient.shared
    private var cancellables = Set<AnyCancellable>()
    
    init() {
        isAuthenticated = authManager.isAuthenticated
        
        if isAuthenticated {
            Task {
                await fetchProfile()
            }
        }
    }
    
    // MARK: - Authentication Methods
    
    func signup(email: String, password: String, displayName: String) async {
        isLoading = true
        errorMessage = nil
        
        do {
            let parameters: [String: Any] = [
                "email": email,
                "password": password,
                "displayName": displayName
            ]
            
            let response: AuthResponse = try await apiClient.request(
                .signup,
                method: .post,
                parameters: parameters
            )
            
            authManager.saveTokens(
                accessToken: response.tokens.accessToken,
                refreshToken: response.tokens.refreshToken
            )
            
            currentUser = response.user
            isAuthenticated = true
            isLoading = false
        } catch {
            errorMessage = error.localizedDescription
            isLoading = false
        }
    }
    
    func login(email: String, password: String) async {
        isLoading = true
        errorMessage = nil
        
        do {
            let parameters: [String: Any] = [
                "email": email,
                "password": password
            ]
            
            let response: AuthResponse = try await apiClient.request(
                .login,
                method: .post,
                parameters: parameters
            )
            
            authManager.saveTokens(
                accessToken: response.tokens.accessToken,
                refreshToken: response.tokens.refreshToken
            )
            
            currentUser = response.user
            isAuthenticated = true
            isLoading = false
        } catch {
            errorMessage = error.localizedDescription
            isLoading = false
        }
    }
    
    func logout() {
        authManager.clearTokens()
        currentUser = nil
        isAuthenticated = false
    }
    
    func fetchProfile() async {
        do {
            let response: ProfileResponse = try await apiClient.request(.profile)
            currentUser = response.user
        } catch {
            // If profile fetch fails, logout
            logout()
        }
    }
    
    func updateProfile(displayName: String?, avatarUrl: String?) async -> Bool {
        isLoading = true
        
        do {
            var parameters: [String: Any] = [:]
            if let displayName = displayName {
                parameters["displayName"] = displayName
            }
            if let avatarUrl = avatarUrl {
                parameters["avatarUrl"] = avatarUrl
            }
            
            let response: ProfileResponse = try await apiClient.request(
                .updateProfile,
                method: .put,
                parameters: parameters
            )
            
            currentUser = response.user
            isLoading = false
            return true
        } catch {
            errorMessage = error.localizedDescription
            isLoading = false
            return false
        }
    }
    
    func changePassword(oldPassword: String, newPassword: String) async -> Bool {
        isLoading = true
        
        do {
            let parameters: [String: Any] = [
                "oldPassword": oldPassword,
                "newPassword": newPassword
            ]
            
            try await apiClient.requestWithoutResponse(
                .changePassword,
                method: .put,
                parameters: parameters
            )
            
            isLoading = false
            return true
        } catch {
            errorMessage = error.localizedDescription
            isLoading = false
            return false
        }
    }
    
    func deleteAccount() async -> Bool {
        isLoading = true
        
        do {
            try await apiClient.requestWithoutResponse(
                .deleteAccount,
                method: .delete
            )
            
            logout()
            isLoading = false
            return true
        } catch {
            errorMessage = error.localizedDescription
            isLoading = false
            return false
        }
    }
}
